<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект-2: Анализ вакансий из HeadHunter
   

In [4]:
import pandas as pd
import psycopg2

In [6]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [17]:
# текст запроса
query_3_1 = f'''
SELECT
     COUNT(*)
FROM
     vacancies
'''

In [18]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_28504\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [13]:
# текст запроса
query_3_2 = f'''
SELECT
     COUNT(*)
FROM
     employers
'''

In [14]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_28504\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
SELECT
     COUNT(*)
FROM
     areas
'''

In [9]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
SELECT
     COUNT(*)
FROM
     industries
'''

In [11]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

Из предварительного анализа наших данных мы можем увидеть, что среди более чем `49 тысяч вакансий` у нас имеются свыше `23 тысяч работодателей` в `1362 регионах` с `294 сферами деятельности`.

Такое количество вакансий, работодателей, уникальных регионов и разнообразных сфер деятельности работодателей говорит о том, что:

* Рынок труда имеет `положительную тенденцию` по отношению к сфере IT, ввиду явной `заинтересованности самих работодателей` в увеличении штата IT сотрудников.

Далее мы проведём детальный анализ вакансий, чтобы узнать следующие подробности:

* В каких регионах вакансий много;

* Как часто работодатели заполняют хотя бы одно из двух полей с зарплатой;

* Разберём значения для нижней и верхней границы зарплатной вилки;

* Какое количество вакансий имеется для каждого сочетания типа рабочего графика и типа трудоустройства, а так же требуемого опыта работы.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [21]:
# текст запроса
query_4_1 = f'''
SELECT
     a.name,
     COUNT(*) cnt
FROM
     areas a
JOIN vacancies v ON v.area_id = a.id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 5
'''

In [22]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_28504\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = f'''
SELECT
     COUNT(*)
FROM
     vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [16]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3 = f'''
SELECT
     AVG(salary_from) avg_salary_from,
     AVG(salary_to) avg_salary_to
FROM
     vacancies
'''

In [18]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0     71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.

In [31]:
# текст запроса
query_4_4 = f'''
SELECT
     schedule,
     employment,
     COUNT(*)
FROM
     vacancies
GROUP BY 1,2
ORDER BY 3 DESC
LIMIT 1 OFFSET 1
'''

In [32]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_28504\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule        employment  count
0  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_4_4 = f'''
SELECT
     experience,
     COUNT(*)
FROM
     vacancies
GROUP BY 1
ORDER BY 2
'''

In [22]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Подводя итог детальному анализу вакансий, можно выделить следующие моменты:

1. `Пятёрку лидеров` по количеству вакансий занимают следующие города:

   * Москва — Санкт-Петербург — Минск — Новосибирск — Алматы
  
    Разница между столицей (`Москва`) и остальными регионами обуславливается количеством населения (Более чем в 2 раза по сравнению с Санкт-Петербургом), что в свою очередь увеличивает количество потенциальных работодателей и количество вакансий в регионе.

2. `У менее, чем половины` предлагаемых вакансий не указана верхняя и нижняя граница заработной платы. Такое решение работодателей во многом зависит от уровня кандидата — если опытного кандидата в процессе переговоров что-то не устроит, то придётся сделать предложение человеку с меньшим опытом. Разумеется, увидев вилку зарплат, он будет рассчитывать на максимум, и получив предложение с гораздо меньшим окладом, разочаруется и тоже может отказаться.

3. `Средние значения` для нижней и верхней границы зарплатной вилки в размере 71065 и 110537 тысяч рублей соответственно характеризуются спросом на квалифицированных специалистов, что становятся всё более ценными. Следовательно компании, готовые конкурировать за лучших сотрудников готовы предлагать и более высокие границы зарплатной вилки, чтобы привлечь и удержать таланты.

4. Помимо очевидного первенства сочетания `Полный день — Полная занятость` (Более 70%) в типах рабочего графика и трудоустройства, стоит также отметить и связку `Удалённая работа — Полная занятость`, что находится на втором месте. Её популярность по сравнению с остальными связками характеризуется возможностью осуществлять профессиональную деятельность сотрудника из любого уголка мира. Ведь сотрудник может работать где угодно, если это место предполагает наличие стабильного интернета и электросети. А значит и вакансия, предлагаемая работодателем располагает к выполнению задач удалённо, без необходимости в постоянном посещении офиса.

5. Среди `Опыта работы` можно выделить заинтересованность работодателей в кандидатах с опытом работы `От 1 года до 3 лет`. Такой интерес можем быть вызван следующими факторами:

   1. Переобучить чужую `"звезду"` может быть долго, дорого и некогда, в отличие от новичка которого всегда можно с лёгкостью переобучить или доучить, а деньги, которые компания вложила в обучение для `"выращивания"` собственных
  кадров, окупятся с большей вероятностью.

   2. Сотрудники с меньшим опытом ожидаемо задают множество вопросов. Они не знают всех тонкостей работы, с энтузиазмом пробуют новое и ошибаются чаще других. Правильный новичок жаждет знаний и ищет любые способы поскорее повысить квалификацию, что также является плюсом как для самого сотрудника, так и работодателя.

Теперь перейдём к анализу самих нанимателей. Посмотрим на лидеров по количеству вакансий, регионы в которых они публикуют свои предложения и их сферы деятельности.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
# текст запроса
query_5_1 = f'''
(SELECT
     e.name
FROM
     employers e
JOIN vacancies v on v.employer_id = e.id
GROUP BY 1
ORDER BY COUNT(*) DESC
LIMIT 1)
UNION
(SELECT
     e.name
FROM
     employers e
JOIN vacancies v on v.employer_id = e.id
GROUP BY 1
ORDER BY COUNT(*) DESC
LIMIT 1 OFFSET 4)
'''

In [25]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name
0  Газпром нефть
1         Яндекс

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [35]:
# текст запроса
query_5_2 = f'''
SELECT
     a.name region_name,
     COUNT(DISTINCT e.id) employers_cnt,
     COUNT(DISTINCT v.id) vacancies_cnt
FROM
     areas a
LEFT JOIN employers e on e.area = a.id
LEFT JOIN vacancies v on v.area_id = a.id
WHERE v.id IS NULL
GROUP BY a.id
ORDER BY 2 desc
LIMIT 1
'''

In [36]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_28504\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region_name  employers_cnt  vacancies_cnt
0      Россия            410              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [39]:
# текст запроса
query_5_3 = f'''
SELECT
     e.name employer_name,
     COUNT(DISTINCT v.area_id)
FROM
     vacancies v
JOIN employers e ON e.id = v.employer_id
GROUP BY 1
ORDER BY 2 DESC
'''

In [38]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer_name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [58]:
# текст запроса
query_5_4 = f'''
SELECT
     COUNT(*)
FROM
     employers e
LEFT JOIN employers_industries ei on ei.employer_id = e.id
WHERE ei.industry_id is NULL
'''

In [59]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [68]:
# текст запроса
query_5_5 = f'''
SELECT
     e.name
FROM
     employers e
LEFT JOIN employers_industries ei on ei.employer_id = e.id
WHERE ei.industry_id IS NOT NULL
GROUP BY 1
HAVING COUNT(*) = 4
ORDER BY 1
LIMIT 1 OFFSET 2
'''

In [69]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_28504\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [75]:
# текст запроса
query_5_6 = f'''
SELECT
     COUNT(*)
FROM
     employers_industries ei
JOIN industries i ON i.id = ei.industry_id
WHERE ei.industry_id IS NOT NULL
AND i.name LIKE 'Разработка программного обеспечения'
GROUP BY i.name
'''

In [76]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [93]:
# код для получения списка городов-милионников
import requests 
from bs4 import BeautifulSoup 

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' 
response = requests.get(url)
# дальше работа с содержимым страницы
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
print(page.title.text,'\n')               
tabl = page.find("table", class_="standard")
town = []
tt = tabl.find("tbody") # в таблице выбираем тело таблицы
qq = tt.find_all("tr")  # в теле таблицы выбираем все ряды

for ss in qq:
  qqq = list(ss.find("a"))  # в каждом ряде находим тег "а", преобразуем его в список
  town.append(qqq[0])       # и добавляем содержимое к списку городов
town = town[1:]             # в первом ряду кусок от названия, отсекаем его
town = tuple(town)
print(town)

Города-миллионеры России — Википедия 

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [140]:
# текст запроса
query_5_7 = f'''
WITH towns AS (
SELECT
     a.name,
     COUNT(v.id) cnt
FROM
     vacancies v
JOIN employers e ON e.id = v.employer_id
JOIN areas a ON a.id = v.area_id
WHERE e.name = 'Яндекс'
AND a.name IN {town}
GROUP BY a.name
)

SELECT
     towns.name,
     towns.cnt
FROM
     towns
UNION ALL
SELECT
     'Total', 
     sum(towns.cnt)  
FROM
     towns
ORDER BY 2 ASC
'''

In [139]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_2924\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7            Самара   26.0
8               Уфа   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

Поговорим о том, что нам дал анализ работодателей:

1. `Безусловным лидером` по количеству опубликованных вакансий является компания `"Яндекс"`, что по совместительству входит в список одних из крупнейших IT-компаний в России. Учитывая рост как самой компании, так и её количества сервисов очевидна потребность в увеличении штата сторудников. Она же является и лидером по количеству регионов, в которых компания публикует свои вакансии — 181 регион, охватывая при этом все города-миллионники России.

2. В наших данных имеются работодатели, у которых `не указана сфера деятельности`. У таких компаний меньший шанс попасть в рекомендацию к соискателю что ищет компании по фильтру `"Отрасль компании"` или попасть в подборку тех, что могли бы его заинтересовать. Возможно, такие компании делают акцент на свой департамент *HR*, что самостоятельно откликаются на подходящие по их мнению резюме и уже во время звонка-интервью рассказывают кандидату подробности о компании.

3. `Самой популярной` сферой деятельности среди компаний является `"Разработка программного обеспечения"`. Она же и сама пополуряная инженерная дисциплина IT в принципе, ведь отвечает как за функционирование, так и за сопровождение всех программных продуктов компании.

Заключительным этапом анализа наших данных будет анализ вакансий, имеющих отношение к данным. Разберём их количество, требуемые ключевые навыки для кандидатов и среднюю заработную плату для каждого типа требуемого опыта работы.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [26]:
# текст запроса
query_6_1 = f'''
SELECT
     COUNT(*)
FROM
     vacancies
WHERE name ILIKE '%data%'
OR name ILIKE '%данн%'
'''

In [27]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_1360\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [95]:
# текст запроса
query_6_2 = f'''
WITH ds AS (
SELECT
     *
FROM
     vacancies
WHERE (name ILIKE '%data scientist%'
OR name ILIKE '%data science%'
OR name ILIKE '%исследователь данных%'
OR name ILIKE '%ML%'
OR name ILIKE '%machine learning%'
OR name ILIKE '%машинн%обучен%')
AND name NOT ILIKE '%HTML%'
)

SELECT
     COUNT(*)
FROM
     ds
WHERE name ILIKE '%junior%'
OR experience ILIKE '%Нет опыта%'
OR employment ILIKE '%Стажировка%'
'''

In [96]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_1360\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [91]:
# текст запроса
query_6_3 = f'''
WITH ds AS (
SELECT
     *
FROM
     vacancies
WHERE (name ILIKE '%data scientist%'
OR name ILIKE '%data science%'
OR name ILIKE '%исследователь данных%'
OR name ILIKE '%ML%'
OR name ILIKE '%machine learning%'
OR name ILIKE '%машинн%обучен%')
AND name NOT ILIKE '%HTML%'
)

SELECT
     COUNT(*)
FROM
     ds
WHERE key_skills ILIKE '%SQL%'
OR key_skills ILIKE '%postgres%'
'''

In [92]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_1360\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [97]:
# текст запроса
query_6_4 = f'''
WITH ds AS (
SELECT
     *
FROM
     vacancies
WHERE (name ILIKE '%data scientist%'
OR name ILIKE '%data science%'
OR name ILIKE '%исследователь данных%'
OR name ILIKE '%ML%'
OR name ILIKE '%machine learning%'
OR name ILIKE '%машинн%обучен%')
AND name NOT ILIKE '%HTML%'
)

SELECT
     COUNT(*)
FROM
     ds
WHERE key_skills ILIKE '%Python%'
'''

In [98]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_1360\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [10]:
# текст запроса
query_6_5 = f'''
WITH ds AS (
SELECT
     *
FROM
     vacancies
WHERE (name ILIKE '%data scientist%'
OR name ILIKE '%data science%'
OR name ILIKE '%исследователь данных%'
OR name LIKE '%ML%'
OR name ILIKE '%machine learning%'
OR name ILIKE '%машинн%обучен%')
AND name NOT ILIKE '%HTML%'
)

SELECT
     round(avg(length(key_skills) - length(replace(key_skills, CHR(9), '')) + 1), 2)
FROM
     ds
WHERE key_skills IS NOT NULL
'''

In [11]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_3064\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [20]:
# текст запроса
query_6_6 = f'''
WITH ds AS (
SELECT
     *
FROM
     vacancies
WHERE (name ILIKE '%data scientist%'
OR name ILIKE '%data science%'
OR name ILIKE '%исследователь данных%'
OR name LIKE '%ML%'
OR name ILIKE '%machine learning%'
OR name ILIKE '%машинн%обучен%')
AND name NOT ILIKE '%HTML%'
)

SELECT
     experience,
     ROUND(AVG(COALESCE((salary_to + salary_from) / 2, salary_from, salary_to)), 0)
FROM
     ds
WHERE salary_from IS NOT NULL
OR salary_to IS NOT NULL
GROUP BY 1
'''

In [21]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\mrlev\AppData\Local\Temp\ipykernel_3064\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0       От 3 до 6 лет  243115.0
1  От 1 года до 3 лет  139675.0
2           Нет опыта   74643.0

***

Из предметного анализа наших данных, мы узнали, что:

 1. Из общего количества вакансий в данных только 1771 `имеет отношение к данным`, что может говорить о большой конкуренции среди соискателей. Из них 51 вакансия подойдёт для специалистов уровня *`Junior`*, что ещё раз указывает на возможную конкуренцию среди начинающих специалистов.

 2. В 229 вакансиях в качестве ключевого навыка указан *`SQL`* или *`postgres`*. В 357 указан *`Python`*. Очевидна популярность и явную необходимость вышеперечисленных языков программирования для специалиста *`Data Science`*.
 
    В среднем же ключевых навыков в вакансиях для *`DS`* c учётом вышеперечисленных языков программирования необходимо 6, как указывают работодатели. Компании ожидают от специалиста знаний и опыта, которые помогут выполнять работу эффективнее и быстрее добиваться результата в работе с разными структурами и данными.

 3. Для специалистов *`DS`* наблюдается быстрый карьерный рост — интенсивность роста заработной платы практически в 2 раза в зависимости от опыта работы, от 76 до 243 тысяч рублей за 3 года!

    Это показывает насколько сильно востребованы компаниями высококвалифицированные *`DS`* специалисты.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

In [1]:
# не забываем закрыть соединение после окончания работы
connection.close()

NameError: name 'connection' is not defined